contracts/PionexContract.sol

```diff
@@ -0,0 +1,632 @@
+    function activateFactors() external onlyOwner {
+        require(factorsTimeLock != 0, "LimitOrder: no pending fee factors");
+        require(block.timestamp >= factorsTimeLock, "LimitOrder: fee factors timelocked");
+        factorsTimeLock = 0;
+        makerFeeFactor = pendingMakerFeeFactor;
+        takerFeeFactor = pendingTakerFeeFactor;
+        profitFeeFactor = pendingProfitFeeFactor;
+        pendingMakerFeeFactor = 0;
+        pendingTakerFeeFactor = 0;
+        pendingProfitFeeFactor = 0;
+
+        emit FactorsUpdated(makerFeeFactor, takerFeeFactor, profitFeeFactor);
+    }
+
+    /// @notice Only owner can call
+    /// @param _newFeeCollector The new address of fee collector
+    function setFeeCollector(address _newFeeCollector) external onlyOwner {
+        require(_newFeeCollector != address(0), "LimitOrder: fee collector can not be zero address");
+        feeCollector = _newFeeCollector;
+
+        emit SetFeeCollector(_newFeeCollector);
+    }
+
+    /// @inheritdoc IPionexContract
+    function fillLimitOrderByTrader(
+        PionexContractLibEIP712.Order calldata _order,
+        bytes calldata _orderMakerSig,
+        TraderParams calldata _params,
+        CoordinatorParams calldata _crdParams
+    ) external override onlyUserProxy nonReentrant returns (uint256, uint256) {
+        bytes32 orderHash = getEIP712Hash(PionexContractLibEIP712._getOrderStructHash(_order));
+
+        _validateOrder(_order, orderHash, _orderMakerSig);
+        bytes32 allowFillHash = _validateFillPermission(orderHash, _params.takerTokenAmount, _params.taker, _crdParams);
+        _validateOrderTaker(_order, _params.taker);
+
+        // Check provided takerToken/makerToken ratio is better than or equal to maker's specfied takerToken/makerToken ratio
+        // -> _params.takerTokenAmount/_params.makerTokenAmount >= _order.takerTokenAmount/_order.makerTokenAmount
+        require(
+            _params.takerTokenAmount.mul(_order.makerTokenAmount) >= _order.takerTokenAmount.mul(_params.makerTokenAmount),
+            "LimitOrder: taker/maker token ratio not good enough"
+        );
+
+        {
+            PionexContractLibEIP712.Fill memory fill = PionexContractLibEIP712.Fill({
+                orderHash: orderHash,
+                taker: _params.taker,
+                recipient: _params.recipient,
+                makerTokenAmount: _params.makerTokenAmount,
+                takerTokenAmount: _params.takerTokenAmount,
+                takerSalt: _params.salt,
+                expiry: _params.expiry
+            });
+            _validateTraderFill(fill, _params.takerSig);
+        }
+
+        (uint256 makerTokenAmount, uint256 takerTokenAmount, uint256 remainingAmount) = _quoteOrderFromMakerToken(_order, orderHash, _params.makerTokenAmount);
+        // Adjust takerTokenAmount according to the provided takerToken/makerToken ratio
+        takerTokenAmount = makerTokenAmount.mul(_params.takerTokenAmount).div(_params.makerTokenAmount);
```

Comment:
`takerTokenAmount` would be derived by exchange ratio, maybe we can skip `takerTokenAmount` calculation in `_quoteOrderFromMakerToken` to save some gas?

Reply:
Fixed in 7da8bba